File: php/features.php
<?php
declare(strict_types=1);

// Intercom SDK feature factory

require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/feature/DebugFeature.php';
require_once __DIR__ . '/feature/IdempotencyFeature.php';
require_once __DIR__ . '/feature/MetricsFeature.php';
require_once __DIR__ . '/feature/PagingFeature.php';
require_once __DIR__ . '/feature/RatelimitFeature.php';
require_once __DIR__ . '/feature/RetryFeature.php';
require_once __DIR__ . '/feature/TestFeature.php';
require_once __DIR__ . '/feature/TimeoutFeature.php';


class IntercomFeatures
{
    public static function make_feature(string $name)
    {
        switch ($name) {
            case "base":
                return new IntercomBaseFeature();
            case "debug":
                return new IntercomDebugFeature();
            case "idempotency":
                return new IntercomIdempotencyFeature();
            case "metrics":
                return new IntercomMetricsFeature();
            case "paging":
                return new IntercomPagingFeature();
            case "ratelimit":
                return new IntercomRatelimitFeature();
            case "retry":
                return new IntercomRetryFeature();
            case "test":
                return new IntercomTestFeature();
            case "timeout":
                return new IntercomTimeoutFeature();
            default:
                return new IntercomBaseFeature();
        }
    }

    /**
     * Does a generated feature class back this name? False for a name only
     * an options extend instance can supply (the station adopt path) - the
     * constructor uses this to skip make_feature for such names instead of
     * adding a stray BaseFeature.
     */
    public static function has_feature(string $name): bool
    {
        switch ($name) {
            case "base":
            case "debug":
            case "idempotency":
            case "metrics":
            case "paging":
            case "ratelimit":
            case "retry":
            case "test":
            case "timeout":
                return true;
            default:
                return false;
        }
    }
}

File: py/intercom_sdk/features.py
# Intercom SDK feature factory

from intercom_sdk.feature.base_feature import IntercomBaseFeature
from intercom_sdk.feature.debug_feature import IntercomDebugFeature
from intercom_sdk.feature.idempotency_feature import IntercomIdempotencyFeature
from intercom_sdk.feature.metrics_feature import IntercomMetricsFeature
from intercom_sdk.feature.paging_feature import IntercomPagingFeature
from intercom_sdk.feature.ratelimit_feature import IntercomRatelimitFeature
from intercom_sdk.feature.retry_feature import IntercomRetryFeature
from intercom_sdk.feature.test_feature import IntercomTestFeature
from intercom_sdk.feature.timeout_feature import IntercomTimeoutFeature


_FEATURES = {
    "base": lambda: IntercomBaseFeature(),
    "debug": lambda: IntercomDebugFeature(),
    "idempotency": lambda: IntercomIdempotencyFeature(),
    "metrics": lambda: IntercomMetricsFeature(),
    "paging": lambda: IntercomPagingFeature(),
    "ratelimit": lambda: IntercomRatelimitFeature(),
    "retry": lambda: IntercomRetryFeature(),
    "test": lambda: IntercomTestFeature(),
    "timeout": lambda: IntercomTimeoutFeature(),
}


def _make_feature(name):
    factory = _FEATURES.get(name)
    if factory is not None:
        return factory()
    return _FEATURES["base"]()


# True when this SDK was generated with the named feature class - the
# constructor's tolerance for extend-carried features reads this (an
# active name with no generated class must not become a BaseFeature
# stray when an extend instance carries it).
def _has_feature(name):
    return name in _FEATURES
